geps/gep-2649/index.md

```diff
@@ -0,0 +1,714 @@
+It is important to note that not every implementation will be able to support
+policy attachment to each resource described in the hierarchy above. When that
+is the case, implementations MUST clearly document which resources a policy may
+be attached to.
+
+#### Attaching Policy to GatewayClass
+GatewayClass may be the trickiest resource to attach policy to. Policy
+attachment relies on the policy being defined within the same scope as the
+target. This ensures that only users with write access to a policy resource in a
+given scope will be able to modify policy at that level. Since GatewayClass is a
+cluster scoped resource, this means that any policy attached to it must also be
+cluster scoped.
+
+GatewayClass parameters provide an alternative to policy attachment that may be
+easier for some implementations to support. These parameters can similarly be
+used to set defaults and requirements for an entire GatewayClass.
+
+### Merging into existing `spec` fields
+
+It's possible (even likely) that configuration in a Policy may need to be merged
+into an existing object's fields somehow, particularly for Inherited policies.
+
+When merging into an existing fields inside an object, Policy objects should
+merge values at a scalar level, not at a struct or object level.
+
+For example, in the `CDNCachingPolicy` example above, the `cdn` struct contains
+a `cachePolicy` struct that contains fields. If an implementation was merging
+this configuration into an existing object that contained the same fields, it
+should merge the fields at a scalar level, with the `includeHost`,
+`includeProtocol`, and `includeQueryString` values being defaulted if they were
+not specified in the object being controlled. Similarly, for `overrides`, the
+values of the innermost scalar fields should overwrite the scalar fields in the
+affected object.
+
+Implementations should not copy any structs from the Policy object directly into the
+affected object, any fields that _are_ overridden should be overridden on a per-field
+basis.
+
+In the case that the field in the Policy affects a struct that is a member of a list,
+each existing item in the list in the affected object should have each of its
+fields compared to the corresponding fields in the Policy.
+
+For non-scalar field _values_, like a list of strings, or a `map[string]string`
+value, the _entire value_ must be overwritten by the value from the Policy. No
+merging should take place. This mainly applies to `overrides`, since for
+`defaults`, there should be no value present in a field on the final object.
+
+This table shows how this works for various types:
+
+|Type|Object config|Override Policy config|Result|
+|----|-------------|----------------------|------|
+|string| `key: "foo"` | `key: "bar"`  | `key: "bar"` |
+|list| `key: ["a","b"]` | `key: ["c","d"]` | `key: ["c","d"]` |
+|`map[string]string`| `key: {"foo": "a", "bar": "b"}` | `key: {"foo": "c", "bar": "d"}` | `key: {"foo": "c", "bar": "d"}` |
```

Comment:
It would be interesting to see an example with lists of structs being affected.

E.g.

|Type|Object config|Override Policy config|Result|
|----|-------------|----------------------|------|
|list| `key: [{"foo": "a", "bar": "b"}, {"foo": "c", "bar": "d"}]`| `key: [{"foo": "a", "baz": "e"}]` | ? |

What should the result be?

Note the second field of the struct in the policy says `baz`, not `bar`.

As for the value in the first field, should we expect `"foo": "c"` in the second struct to be replaced with `"foo": "a"`?

I wonder if it would make a difference if the field was called `name` instead of `foo`. Imagine `listType=map, listMapKey=name`.

Reply:
This is a good point, I'll add some rules around listMapType objects (since they are effectively maps).

Reply:
Thanks, @youngnick! Awesome example, to leave no doubts!